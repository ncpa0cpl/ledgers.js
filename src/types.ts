import type { Entity } from "./entity/entity";
import type { DeepPartial, ReadonlyKeys, Serial } from "./type-utils";

export type EntityData<T extends object> = Omit<
  Serial<T>,
  "name" | "createdAt" | "updatedAt" | ReadonlyKeys<T>
> & { id?: string };

export type EntityChangeData<T extends object> = DeepPartial<
  Omit<EntityData<T>, "id">
>;

export enum EventType {
  CREATE = "CREATE",
  CHANGE = "CHANGE",
  BREAKPOINT = "BREAKPOINT",
}

export class Copy {
  id!: string;
}

export type EventData<T extends object> = {
  id: string;
  timestamp: number;
  type: EventType;
  data: EntityChangeData<T> | EntityData<T>;
  breakpoint?: string | number;
};

export type SerializedEntityListEvents<T extends object> = [
  string,
  EventData<T>[]
][];

export type SerializedBreakpoints = {
  ledgerBreakpoints: Array<string | number>;
};

export type SerializedEntities = {
  singletonEntities: Record<string, EventData<any>[]>;
  listEntities: Record<string, SerializedEntityListEvents<any>>;
  copies: Record<string, Copy[]>;
};

export type SerializedLedger = SerializedEntities &
  SerializedBreakpoints & {
    name: string;
  };

export enum EntityReferenceType {
  SINGLETON = "SINGLETON",
  LIST = "LIST",
  COPY = "COPY",
}

export type Reference<E extends Entity | Copy> = {
  ledgerName: string;
  name: string;
  type: EntityReferenceType;
  id: E["id"];
};
