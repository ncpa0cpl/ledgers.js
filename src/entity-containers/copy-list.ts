import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import { Ledger } from "../ledger/ledger";
import type { TransactionInterface } from "../ledger/transaction";
import type { Copy } from "../types";

export class CopyList<C extends Copy> {
  /** @internal */
  static _loadFrom<D extends Copy>(copyList: CopyList<any>, copies: D[]) {
    if (copyList.committed.size > 0) {
      throw new LedgerError(ErrorCode.DESERIALIZING_ON_NON_EMPTY_LEDGER);
    }

    for (const c of copies) {
      copyList.put(c);
    }
  }

  /** @internal */
  static _serialize<C extends Copy>(list: CopyList<C>): C[] {
    if (list.isTransactionPending) {
      throw new LedgerError(ErrorCode.SERIALIZING_DURING_TRANSACTION);
    }

    return list.getAll();
  }

  private parentLedger: Ledger;
  private name: string;
  private committed = new Map<string, C>();
  private staged = new Map<string, C>();
  private txInterface: TransactionInterface;

  constructor(parentLedger: Ledger, name: string) {
    this.parentLedger = parentLedger;
    this.name = name;

    Ledger._getEntityController(parentLedger).registerCopyList(this, this.name);

    this.txInterface = {
      commit: () => this.commit(),
      rollback: () => this.rollback(),
    };
  }

  private mergeMaps() {
    const tmpCopies = new Map(this.committed);

    for (const [name, copy] of this.staged) {
      tmpCopies.set(name, copy);
    }

    return tmpCopies;
  }

  private get isTransactionPending(): boolean {
    return this.staged.size > 0;
  }

  private commit() {
    this.committed = this.mergeMaps();
    this.staged = new Map();
  }

  private rollback() {
    this.staged.clear();
  }

  private addToTransaction(): void {
    const tx = Ledger._getTransaction(this.parentLedger);

    if (tx) {
      tx.add(this.txInterface);
    } else {
      this.commit();
    }
  }

  has(id: string): boolean {
    return this.committed.has(id) || this.staged.has(id);
  }

  put(copy: C): void {
    this.staged.set(copy.id, copy);
    this.addToTransaction();
  }

  get(id: string): C | undefined {
    return this.staged.get(id) ?? this.committed.get(id);
  }

  getAll(): C[] {
    return [...this.mergeMaps().values()];
  }
}
