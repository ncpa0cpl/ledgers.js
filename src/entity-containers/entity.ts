import { BaseEntity } from "../entity/base-entity";
import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import type { GenerateEventData } from "../events/event";
import { Event } from "../events/event";
import { EventList } from "../events/event-list";
import { Ledger } from "../ledger/ledger";
import type {
  AdditionalEventData,
  EntityChangeData,
  EntityData,
  SerializedEvent,
} from "../types";
import { extractEntityIdFromEvent } from "../utilities/extract-entity-id-from-event";

export class Entity<E extends BaseEntity> {
  /** @internal */
  static _loadFrom<E2 extends BaseEntity>(
    singleton: Entity<E2>,
    eventData: SerializedEvent[],
  ): void {
    if (singleton.events.length > 0) {
      throw new LedgerError(ErrorCode.DESERIALIZING_ON_NON_EMPTY_LEDGER);
    }

    const migrationController = Ledger._getMigrationController(
      singleton.parentLedger,
    );

    for (const e of eventData) {
      if (singleton.entityName !== e.metadata.entity) {
        throw new LedgerError(ErrorCode.EVENT_ASSOCIATION_ERROR);
      }

      const event = migrationController.migrateEvent(Event._loadFrom(e));

      singleton.events.add(event);
    }

    singleton.events.commit();
  }

  /** @internal */
  static _serialize<E extends BaseEntity>(
    singleton: Entity<E>,
  ): SerializedEvent[] {
    return singleton.events.serialize();
  }

  /** @internal */
  static _addBreakpointEvent<E extends BaseEntity>(
    entity: Entity<E>,
    breakpoint: string | number,
  ): void {
    return entity.eventBreakpoint(breakpoint);
  }

  private readonly parentLedger: Ledger;
  private readonly entityName: string;
  private readonly entityConstructor: new (parent: Ledger) => E;
  private events: EventList<E>;

  constructor(ledger: Ledger, entityConstructor: new (parent: Ledger) => E) {
    this.parentLedger = ledger;
    this.entityConstructor = entityConstructor;
    const tmpEntity = new entityConstructor(ledger);
    this.entityName = tmpEntity.name;
    this.events = new EventList<E>(this.parentLedger);

    if (!this.entityName) {
      throw new LedgerError(ErrorCode.ENTITY_NAME_NOT_SPECIFIED);
    }

    Ledger._getEntityController(ledger).registerSingleton(this);
  }

  private addToTransaction(): void {
    const tx = Ledger._getTransaction(this.parentLedger);

    if (tx) {
      tx.add(this.events);
    } else {
      this.events.commit();
    }
  }

  private eventBreakpoint(
    breakpoint: string | number,
    eventMetadata?: AdditionalEventData,
  ): void {
    if (this.events.length === 0) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    const event = Event._generateBreakpointEvent<E>(
      this.parentLedger,
      breakpoint,
      meta,
    );

    this.events.add(event);

    this.addToTransaction();
  }

  create(initData: EntityData<E>, eventMetadata?: AdditionalEventData): string {
    if (this.events.length > 0) {
      throw new LedgerError(ErrorCode.ENTITY_ALREADY_CREATED);
    }

    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    initData.id ??= this.parentLedger.generateNextID();

    const event = Event._generateCreateEvent(this.parentLedger, initData, meta);

    this.events.add(event);

    this.addToTransaction();

    return initData.id;
  }

  change(
    changes: EntityChangeData<E>,
    eventMetadata?: AdditionalEventData,
  ): void {
    if (this.events.length === 0) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    const event = Event._generateChangeEvent<E>(
      this.parentLedger,
      changes,
      meta,
    );

    this.events.add(event);

    this.addToTransaction();
  }

  getName(): string {
    return this.entityName;
  }

  isInitiated(): boolean {
    return this.events.length !== 0;
  }

  getID(): string {
    const eventList = this.events.getAsArray();
    const firstEvent = eventList[0];

    if (firstEvent === undefined) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    return extractEntityIdFromEvent(firstEvent);
  }

  get(breakpoint?: string | number): E {
    if (this.events.length === 0) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    if (
      breakpoint !== undefined &&
      !this.events.hasCreateEventBeforeBreakpoint(breakpoint)
    ) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    const entity = new this.entityConstructor(this.parentLedger);

    BaseEntity._applyEvents(entity, this.events.getAsArray(breakpoint));

    return entity;
  }
}
