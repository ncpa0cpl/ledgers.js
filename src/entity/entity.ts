import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import type { Event } from "../events/event";

export abstract class Entity {
  abstract readonly name: string;
  readonly id!: string;

  createdAt!: number;
  updatedAt!: number;

  static _applyEvents<T extends Entity>(entity: T, events: Event<T>[]): void {
    if (events.length === 0) {
      throw new LedgerError(ErrorCode.EMPTY_EVENTS_LIST);
    }

    for (const event of events) {
      event.apply(entity);
    }

    entity.createdAt = events[0]!.timestamp;
    entity.updatedAt = events[events.length - 1]!.timestamp;
  }
}
