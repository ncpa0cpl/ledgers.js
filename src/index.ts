export { CopiesList as CopyList } from "./entity-containers/copy-list";
export { Entity as EntitySingleton } from "./entity-containers/entity";
export { EntityList } from "./entity-containers/entity-list";
export { BaseEntity as Entity } from "./entity/base-entity";
export { Ledger } from "./ledger/ledger";
export type {
  Copy,
  EntityChangeData,
  EntityData,
  EntityReferenceType,
  Reference,
  SerializedEntities,
  SerializedEntityListEvents,
  SerializedLedger,
} from "./types";
