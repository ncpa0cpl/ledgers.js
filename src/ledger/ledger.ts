import * as uuid from "uuid";
import type { Entity } from "../entity/entity";
import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import type { Copy, Reference, SerializedLedger } from "../types";
import { EntityReferenceType } from "../types";
import { BreakpointController } from "./breakpoint-controller";
import { EntitiesController } from "./entities-controller";
import { Transaction } from "./transaction";

type InitializeArgs<T> = T extends new (...args: infer A) => any ? A : never;

export abstract class Ledger {
  static loadFrom<T extends new (...args: any[]) => Ledger>(
    this: T,
    data: SerializedLedger,
    ...args: InitializeArgs<T>
  ): InstanceType<T> {
    const ledger = new this(...args);

    if (data.name !== ledger.name) {
      throw new LedgerError(ErrorCode.LEDGER_NAMES_DO_NOT_MATCH);
    }

    ledger.entities.loadFrom(data);
    ledger.breakpoints.loadFrom(data);

    return ledger as InstanceType<T>;
  }

  /** @internal */
  static _getTransaction(l: Ledger) {
    return l.transaction;
  }

  /** @internal */
  static _getEntityController(l: Ledger): EntitiesController {
    return l.entities;
  }

  /** @internal */
  static _getBreakpointController(l: Ledger): BreakpointController {
    return l.breakpoints;
  }

  private entities = new EntitiesController();
  private breakpoints = new BreakpointController(this);
  private transaction?: Transaction;

  abstract name: string;

  generateNextID(): string {
    return uuid.v4();
  }

  generateTimestamp(): number {
    return Date.now();
  }

  createReference<T extends Entity | Copy>(entity: T): Reference<T> {
    let type: EntityReferenceType = EntityReferenceType.SINGLETON;
    const isSingleton = !!this.entities.findSingletonEntity(entity.id);

    if (!isSingleton) {
      type = EntityReferenceType.LIST;
      const isList = !!this.entities.findListEntity(entity.id);

      if (!isList) {
        type = EntityReferenceType.COPY;
        const isCopy = !!this.entities.findCopy(entity.id);

        if (!isCopy) {
          throw new LedgerError(ErrorCode.ENTITY_NOT_FOUND);
        }
      }
    }

    let name: string;

    if (type === EntityReferenceType.COPY) {
      name = this.entities.getCopyName(entity.id);
    } else {
      name = (entity as Entity).name;
    }

    return {
      id: entity.id,
      ledgerName: this.name,
      name,
      type,
    };
  }

  resolveReference<T extends Entity | Copy>(ref: Reference<T>): T {
    if (ref.ledgerName !== this.name) {
      throw new LedgerError(ErrorCode.LEDGER_NAMES_DO_NOT_MATCH);
    }

    switch (ref.type) {
      case EntityReferenceType.SINGLETON: {
        const singleton = this.entities.getSingletonByName(ref.name);

        if (!singleton.isInitiated() || singleton.getID() !== ref.id) {
          throw new LedgerError(ErrorCode.ENTITY_NOT_FOUND);
        }

        return singleton.get() as T;
      }
      case EntityReferenceType.LIST: {
        const list = this.entities.getListByName(ref.name);

        return list.get(ref.id) as T;
      }
      case EntityReferenceType.COPY: {
        const copyList = this.entities.getCopiesByName(ref.name);

        const copy = copyList.get(ref.id);

        if (!copy) {
          throw new LedgerError(ErrorCode.ENTITY_NOT_FOUND);
        }

        return copy as T;
      }
      default:
        throw new Error();
    }
  }

  startTransaction(): void {
    if (this.transaction) {
      throw new LedgerError(ErrorCode.TRANSACTION_ALREADY_IN_PROGRESS);
    }

    this.transaction = new Transaction();
  }

  commitTransaction(): void {
    this.transaction?.commit();

    this.transaction = undefined;
  }

  rollbackTransaction(): void {
    this.transaction?.rollback();

    this.transaction = undefined;
  }

  tx(callback: () => void): void {
    this.startTransaction();

    try {
      callback();
      this.commitTransaction();
    } catch (e) {
      this.rollbackTransaction();
      throw e;
    }
  }

  addBreakpoint(breakpoint: string | number): void {
    this.breakpoints.addBreakpoint(breakpoint);
  }

  hasBreakpoint(breakpoint: string | number): boolean {
    return this.breakpoints.hasBreakpoint(breakpoint);
  }

  getExistingBreakpoints(): Array<string | number> {
    return this.breakpoints.getBreakpoints();
  }

  serialize(): SerializedLedger {
    if (this.transaction) {
      throw new LedgerError(ErrorCode.SERIALIZING_DURING_TRANSACTION);
    }

    return {
      name: this.name,
      ...this.entities.serialize(),
      ...this.breakpoints.serialize(),
    };
  }

  getSnapshot(breakpoint?: string | number): object {
    return { ...this.entities.getSnapshot(breakpoint) };
  }

  getHistory(): Array<{
    time: number;
    breakpoint?: string | number;
    snapshot: object;
  }> {
    return [
      ...this.breakpoints.getBreakpointsWithTimestamps().map((breakpoint) => ({
        time: breakpoint.createdAt,
        breakpoint: breakpoint.breakpointID,
        snapshot: this.getSnapshot(breakpoint.breakpointID),
      })),
      {
        time: this.generateTimestamp(),
        snapshot: this.getSnapshot(),
      },
    ];
  }
}
