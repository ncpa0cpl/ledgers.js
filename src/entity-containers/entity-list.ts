import { BaseEntity } from "../entity/base-entity";
import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import type { GenerateEventData } from "../events/event";
import { Event } from "../events/event";
import { EventList } from "../events/event-list";
import { Ledger } from "../ledger/ledger";
import type {
  AdditionalEventData,
  EntityChangeData,
  EntityData,
  SerializedEntityListEvents,
  SerializedEvent,
} from "../types";

export class EntityList<E extends BaseEntity> {
  /** @internal */
  static _loadFrom<E2 extends BaseEntity>(
    list: EntityList<E2>,
    eventData: SerializedEntityListEvents,
  ): void {
    if (list.entitiesEvents.size > 0) {
      throw new LedgerError(ErrorCode.DESERIALIZING_ON_NON_EMPTY_LEDGER);
    }

    const migrationController = Ledger._getMigrationController(
      list.parentLedger,
    );

    for (const [id, events] of eventData) {
      const eventList = new EventList<E2>(list.parentLedger);

      for (const e of events) {
        if (list.entityName !== e.metadata.entity) {
          throw new LedgerError(ErrorCode.EVENT_ASSOCIATION_ERROR);
        }

        const event = migrationController.migrateEvent(Event._loadFrom(e));

        eventList.add(event);
      }

      eventList.commit();

      list.entitiesEvents.set(id, eventList);
    }
  }

  /** @internal */
  static _serialize<E extends BaseEntity>(
    list: EntityList<E>,
  ): SerializedEntityListEvents {
    for (const [, eventList] of list.entitiesEvents) {
      if (eventList.isTransactionPending) {
        throw new LedgerError(ErrorCode.SERIALIZING_DURING_TRANSACTION);
      }
    }

    return [...list.entitiesEvents.entries()].map(
      ([id, events]): [string, SerializedEvent[]] => [id, events.serialize()],
    );
  }

  /** @internal */
  static _addBreakpointEvent<E extends BaseEntity>(
    list: EntityList<E>,
    breakpoint: string | number,
  ): void {
    return list.eventBreakpoint(breakpoint);
  }

  private readonly parentLedger: Ledger;
  private readonly entityName: string;
  private readonly entityConstructor: new (parent: Ledger) => E;
  private readonly entitiesEvents = new Map<string, EventList<E>>();

  constructor(ledger: Ledger, entityConstructor: new (parent: Ledger) => E) {
    this.parentLedger = ledger;
    this.entityConstructor = entityConstructor;
    this.entityName = new entityConstructor(ledger).name;

    if (!this.entityName) {
      throw new LedgerError(ErrorCode.ENTITY_NAME_NOT_SPECIFIED);
    }

    Ledger._getEntityController(ledger).registerList(this);
  }

  private postTransaction() {
    const deleteEntries: string[] = [];

    for (const [id, eventList] of this.entitiesEvents) {
      if (eventList.length === 0) deleteEntries.push(id);
    }

    for (const id of deleteEntries) {
      this.entitiesEvents.delete(id);
    }
  }

  private addToTransaction(eventList: EventList<E>): void {
    const tx = Ledger._getTransaction(this.parentLedger);

    if (tx) {
      tx.add(eventList);
      tx.addPostTxHook(() => this.postTransaction());
    } else {
      eventList.commit();
      this.postTransaction();
    }
  }

  private createEntityFromEvents(
    events: EventList<E>,
    breakpoint?: string | number,
  ): E {
    if (events.length === 0) {
      throw new LedgerError(ErrorCode.EMPTY_EVENTS_LIST);
    }

    const entity = new this.entityConstructor(this.parentLedger);

    BaseEntity._applyEvents(entity, events.getAsArray(breakpoint));

    return entity;
  }

  private eventBreakpoint(
    breakpoint: string | number,
    eventMetadata?: AdditionalEventData,
  ): void {
    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    for (const eventList of this.entitiesEvents.values()) {
      const event = Event._generateBreakpointEvent<E>(
        this.parentLedger,
        breakpoint,
        meta,
      );

      eventList.add(event);
      this.addToTransaction(eventList);
    }
  }

  create(initData: EntityData<E>, eventMetadata?: AdditionalEventData): string {
    initData.id ??= this.parentLedger.generateNextID();

    if (this.entitiesEvents.has(initData.id)) {
      throw new LedgerError(ErrorCode.DUPLICATE_IDENTIFIER);
    }

    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    const previousBreakpoints = Ledger._getBreakpointController(
      this.parentLedger,
    ).getBreakpoints();

    const eventList = new EventList<E>(this.parentLedger);

    for (const breakpoint of previousBreakpoints) {
      eventList.add(
        Event._generateBreakpointEvent<E>(this.parentLedger, breakpoint, meta),
      );
    }

    const event = Event._generateCreateEvent(this.parentLedger, initData, meta);

    eventList.add(event);

    this.entitiesEvents.set(initData.id, eventList);

    this.addToTransaction(eventList);

    return initData.id;
  }

  change(
    id: string,
    changes: EntityChangeData<E>,
    eventMetadata?: AdditionalEventData,
  ): void {
    const eventList = this.entitiesEvents.get(id);

    if (eventList === undefined) {
      throw new LedgerError(ErrorCode.UNKNOWN_IDENTIFIER);
    }

    const meta: GenerateEventData = {
      ...eventMetadata,
      entity: this.entityName,
    };

    const event = Event._generateChangeEvent<E>(
      this.parentLedger,
      changes,
      meta,
    );

    eventList.add(event);

    this.addToTransaction(eventList);
  }

  getName(): string {
    return this.entityName;
  }

  getAll(breakpoint?: string | number): E[] {
    const entities: E[] = [];

    for (const events of this.entitiesEvents.values()) {
      if (
        breakpoint !== undefined &&
        !events.hasCreateEventBeforeBreakpoint(breakpoint)
      )
        continue;

      entities.push(this.createEntityFromEvents(events, breakpoint));
    }

    return entities;
  }

  get(id: string, breakpoint?: string | number): E {
    const eventList = this.entitiesEvents.get(id);

    if (!eventList) {
      throw new LedgerError(ErrorCode.UNKNOWN_IDENTIFIER);
    }

    if (
      breakpoint !== undefined &&
      !eventList.hasCreateEventBeforeBreakpoint(breakpoint)
    ) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    return this.createEntityFromEvents(eventList, breakpoint);
  }

  has(id: string): boolean {
    return this.entitiesEvents.has(id);
  }
}
