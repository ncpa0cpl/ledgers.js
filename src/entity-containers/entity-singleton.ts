import { Entity } from "../entity/entity";
import { ErrorCode } from "../errors/error-codes";
import { LedgerError } from "../errors/ledger-error";
import { Event } from "../events/event";
import { EventList } from "../events/event-list";
import { Ledger } from "../ledger/ledger";
import type { EntityChangeData, EntityData, EventData } from "../types";

export class EntitySingleton<E extends Entity> {
  static _loadFrom<E2 extends Entity>(
    singleton: EntitySingleton<E2>,
    eventData: EventData<any>[]
  ): void {
    if (singleton.events.length > 0) {
      throw new LedgerError(ErrorCode.DESERIALIZING_ON_NON_EMPTY_LEDGER);
    }

    eventData.forEach((e) => singleton.events.add(new Event(e)));
    singleton.events.commit();
  }

  static _serialize<E extends Entity>(
    singleton: EntitySingleton<E>
  ): EventData<E>[] {
    return singleton.events.serialize();
  }

  private readonly parentLedger: Ledger;
  private readonly entityName: string;
  private readonly entityConstructor: new (parent: Ledger) => E;
  private events = new EventList<E>();

  constructor(ledger: Ledger, entityConstructor: new (parent: Ledger) => E) {
    this.parentLedger = ledger;
    this.entityConstructor = entityConstructor;
    this.entityName = new entityConstructor(ledger).name;

    if (!this.entityName) {
      throw new LedgerError(ErrorCode.ENTITY_NAME_NOT_SPECIFIED);
    }

    Ledger._getEntityController(ledger).registerSingleton(this);
  }

  private addToTransaction(): void {
    const tx = Ledger._getTransaction(this.parentLedger);

    if (tx) {
      tx.add(this.events);
    } else {
      this.events.commit();
    }
  }

  eventCreate(initData: EntityData<E>): string {
    if (this.events.length > 0) {
      throw new LedgerError(ErrorCode.ENTITY_ALREADY_CREATED);
    }

    initData.id ??= this.parentLedger.generateNextID();

    const event = Event.generateCreateEvent(this.parentLedger, initData);

    this.events.add(event);

    this.addToTransaction();

    return initData.id;
  }

  eventChange(changes: EntityChangeData<E>): void {
    if (this.events.length === 0) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    const event = Event.generateChangeEvent<E>(this.parentLedger, changes);

    this.events.add(event);

    this.addToTransaction();
  }

  getName(): string {
    return this.entityName;
  }

  getID(): string {
    if (this.events.length === 0) {
      return "";
    }

    return this.events.getAsArray()[0]!.data.id ?? "";
  }

  get(): E {
    if (this.events.length === 0) {
      throw new LedgerError(ErrorCode.ENTITY_NOT_YET_CREATED);
    }

    const entity = new this.entityConstructor(this.parentLedger);

    Entity._applyEvents(entity, this.events.getAsArray());

    return entity;
  }
}
