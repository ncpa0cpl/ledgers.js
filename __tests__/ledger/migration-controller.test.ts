import { EntitySingleton } from "../../src/entity-containers/entity-singleton";
import { Entity } from "../../src/entity/entity";
import { Ledger } from "../../src/ledger/ledger";
import { EventType, MigrationInterface } from "../../src/types";

describe("migrations", () => {
  it("should apply migrations", () => {
    class FooEntityV1 extends Entity {
      name = "FooEntity";

      a: number;

      constructor() {
        super();
      }
    }

    class LedgerV1 extends Ledger {
      static version = 1;
      name = "Ledger";

      foo = new EntitySingleton(this, FooEntityV1);

      constructor() {
        super();
        this.foo.create({ a: 12 });
      }
    }

    class FooEntityV2 extends Entity {
      name = "FooEntity";

      a: number;
      b: string;
    }

    class LedgerV2 extends Ledger {
      static version = 2;
      name = "Ledger";

      foo = new EntitySingleton(this, FooEntityV2);
    }

    class FooEntityV3 extends Entity {
      name = "FooEntity";

      a: string;
      b: string;
    }

    class LedgerV3 extends Ledger {
      static version = 3;
      name = "Ledger";

      foo = new EntitySingleton(this, FooEntityV3);
    }

    const migration1to2: MigrationInterface<FooEntityV1, FooEntityV2> = {
      entity: "FooEntity",
      version: 2,
      migrateCreateEvent(eventData, meta) {
        return {
          ...eventData,
          b: eventData.a.toString(),
        };
      },
    };

    const migration2to3: MigrationInterface<FooEntityV2, FooEntityV3> = {
      entity: "FooEntity",
      version: 3,
      migrateCreateEvent(eventData, meta) {
        return {
          ...eventData,
          a: `Num(${eventData.b})`,
        };
      },
    };

    LedgerV2.registerMigrations(migration1to2);
    LedgerV3.registerMigrations(migration1to2, migration2to3);

    const ledger1 = new LedgerV1();
    const ledger1Serialized = ledger1.serialize();

    ledger1Serialized.singletonEntities[""];

    expect(ledger1Serialized).toMatchObject({
      singletonEntities: expect.objectContaining({
        FooEntity: [
          expect.objectContaining({
            metadata: expect.objectContaining({
              ledgerVersion: 1,
              type: EventType.CREATE,
              entity: "FooEntity",
            }),
            instructions: expect.arrayContaining([
              expect.objectContaining({
                propertyPath: ["id"],
              }),
              expect.objectContaining({
                propertyPath: ["a"],
                value: 12,
              }),
            ]),
          }),
        ],
      }),
    });

    const ledger2 = LedgerV2.loadFrom(ledger1Serialized);
    const ledger2Serialized = ledger2.serialize();

    expect(ledger2Serialized).toMatchObject({
      singletonEntities: expect.objectContaining({
        FooEntity: [
          expect.objectContaining({
            metadata: expect.objectContaining({
              ledgerVersion: 2,
              type: EventType.CREATE,
              entity: "FooEntity",
              appliedMigrations: ["1:2"],
            }),
            instructions: expect.arrayContaining([
              expect.objectContaining({
                propertyPath: ["id"],
              }),
              expect.objectContaining({
                propertyPath: ["a"],
                value: 12,
              }),
              expect.objectContaining({
                propertyPath: ["b"],
                value: "12",
              }),
            ]),
          }),
        ],
      }),
    });

    const ledger3from1 = LedgerV3.loadFrom(ledger1Serialized);
    const ledger3from1Serialized = ledger3from1.serialize();

    const ledger3from2 = LedgerV3.loadFrom(ledger2Serialized);
    const ledger3from2Serialized = ledger3from2.serialize();

    const expected = {
      singletonEntities: expect.objectContaining({
        FooEntity: [
          expect.objectContaining({
            metadata: expect.objectContaining({
              ledgerVersion: 3,
              type: EventType.CREATE,
              entity: "FooEntity",
              appliedMigrations: ["1:2", "2:3"],
            }),
            instructions: expect.arrayContaining([
              expect.objectContaining({
                propertyPath: ["id"],
              }),
              expect.objectContaining({
                propertyPath: ["a"],
                value: "Num(12)",
              }),
              expect.objectContaining({
                propertyPath: ["b"],
                value: "12",
              }),
            ]),
          }),
        ],
      }),
    };

    expect(ledger3from1Serialized).toMatchObject(expected);
    expect(ledger3from2Serialized).toMatchObject(expected);

    expect(ledger3from1.getSnapshot()).toMatchSnapshot();
  });
});
