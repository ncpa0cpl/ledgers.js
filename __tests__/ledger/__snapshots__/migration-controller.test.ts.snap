// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`migrations should apply migrations 1`] = `
Object {
  "copies": Object {},
  "listEntities": Object {},
  "singletonEntities": Object {
    "FooEntity": FooEntityV3 {
      "a": "Num(12)",
      "b": "12",
      "createdAt": 1666032490844,
      "id": "6cab2145-3b2b-4c4e-9c1f-50994e4363fd",
      "name": "FooEntity",
      "updatedAt": 1666032490844,
    },
  },
}
`;
